File: app/services/completed_loan_services/tab_documents.rb
module CompletedLoanServices
  class TabDocuments
    attr_accessor :borrower, :secondary_borrower

    def initialize(args)
      @borrower = args[:borrower]
      @secondary_borrower = args[:secondary_borrower]
    end

    def call
      return borrower_documents_completed? unless secondary_borrower.present?
      borrower_documents_completed? && co_borrower_documents_completed?
    end

    def borrower_documents_completed?
      borrower.documents.where(is_required: true, original_filename: nil).size == 0
    end

    def co_borrower_documents_completed?
      secondary_borrower.documents.where(is_required: true, original_filename: nil).size == 0
    end
  end
end

File: client/assets/javascripts/mixins/CompletedLoanMixins/TabDocuments.js
var TabDocuments = {
  /**
   * Check tab document is completed or not
   * @type {boolean}
   */
  documentsCompleted: function(loan){
    var borrower = loan.borrower;
    var secondaryBorrower = loan.secondary_borrower;

    if(secondaryBorrower === undefined || secondaryBorrower === null)
      return this.borrowerDocumentsCompleted(borrower);

    return (this.borrowerDocumentsCompleted(borrower) && this.borrowerDocumentsCompleted(secondaryBorrower));
  },

  borrowerDocumentsCompleted: function(borrower){
    var noUploadedFiles = _.filter(borrower.documents, function(document){ return document.is_required == true && document.original_filename === null });

    return noUploadedFiles == 0;
  }
}

module.exports = TabDocuments;
